File: src/chatbot.c
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <limits.h>
#include <strings.h>
#include "chatbot.h"

// hash table implementation from here
// https://gist.githubusercontent.com/tonious/1377667/raw/c814d0833c8699dc017871931a5c5bee11af0f64/hash.c

struct entry_s {
  char *key;
  char *value;
  struct entry_s *next;
};

typedef struct entry_s entry_t;

struct hashtable_s {
  int size;
  struct entry_s **table;
};

typedef struct hashtable_s hashtable_t;


/* Create a new hashtable. */
hashtable_t *ht_create( int size ) {

  hashtable_t *hashtable = NULL;
  int i;

  if( size < 1 ) return NULL;

  /* Allocate the table itself. */
  if( ( hashtable = malloc( sizeof( hashtable_t ) ) ) == NULL ) {
    return NULL;
  }

  /* Allocate pointers to the head nodes. */
  if( ( hashtable->table = malloc( sizeof( entry_t * ) * size ) ) == NULL ) {
    return NULL;
  }
  for( i = 0; i < size; i++ ) {
    hashtable->table[i] = NULL;
  }

  hashtable->size = size;

  return hashtable;
}

/* Hash a string for a particular hash table. */
int ht_hash( hashtable_t *hashtable, char *key ) {

  unsigned long int hashval = 0;
  unsigned long i = 0;

  /* Convert our string to an integer */
  while( hashval < ULONG_MAX && i < strlen( key ) ) {
    hashval = hashval << 8;
    hashval += key[ i ];
    i++;
  }

  return hashval % hashtable->size;
}

/* Create a key-value pair. */
entry_t *ht_newpair( char *key, char *value ) {
  entry_t *newpair;

  if( ( newpair = malloc( sizeof( entry_t ) ) ) == NULL ) {
    return NULL;
  }

  if( ( newpair->key = strdup( key ) ) == NULL ) {
    return NULL;
  }

  if( ( newpair->value = strdup( value ) ) == NULL ) {
    return NULL;
  }

  newpair->next = NULL;

  return newpair;
}

/* Insert a key-value pair into a hash table. */
void ht_set( hashtable_t *hashtable, char *key, char *value ) {
  int bin = 0;
  entry_t *newpair = NULL;
  entry_t *next = NULL;
  entry_t *last = NULL;

  bin = ht_hash( hashtable, key );

  next = hashtable->table[ bin ];

  while( next != NULL && next->key != NULL && strcmp( key, next->key ) > 0 ) {
    last = next;
    next = next->next;
  }

  /* There's already a pair.  Let's replace that string. */
  if( next != NULL && next->key != NULL && strcmp( key, next->key ) == 0 ) {

    free( next->value );
    next->value = strdup( value );

  /* Nope, could't find it.  Time to grow a pair. */
  } else {
    newpair = ht_newpair( key, value );

    /* We're at the start of the linked list in this bin. */
    if( next == hashtable->table[ bin ] ) {
      newpair->next = next;
      hashtable->table[ bin ] = newpair;

    /* We're at the end of the linked list in this bin. */
    } else if ( next == NULL ) {
      last->next = newpair;

    /* We're in the middle of the list. */
    } else  {
      newpair->next = next;
      last->next = newpair;
    }
  }
}

/* Retrieve a key-value pair from a hash table. */
char *ht_get( hashtable_t *hashtable, char *key ) {
  int bin = 0;
  entry_t *pair;

  bin = ht_hash( hashtable, key );

  /* Step through the bin, looking for our value. */
  pair = hashtable->table[ bin ];
  while( pair != NULL && pair->key != NULL && strcmp( key, pair->key ) > 0 ) {
    pair = pair->next;
  }

  /* Did we actually find anything? */
  if( pair == NULL || pair->key == NULL || strcmp( key, pair->key ) != 0 ) {
    return NULL;

  } else {
    return pair->value;
  }

}

int main(void) {

  char line[LINELENGTH];
  char *word;
  printf("$ Chatbot v1.0.0!\n");

  hashtable_t *hashtable = ht_create(65536);
  ht_set(hashtable, "hi", "hello");
  ht_set(hashtable, "hey", "hello");
  ht_set(hashtable, "hear", "What you heard is right");
  ht_set(hashtable, "python", "Yo, I love Python");
  ht_set(hashtable, "light", "I like light");
  ht_set(hashtable, "What", "It is clear, ain't it?");

  while(1) {
    printf("\n$ (user) ");
    fgets(line, LINELENGTH, stdin);
    if (strlen(line) <= 1) break; /*exit program*/
    word = strtok(line, SEPCHARS); /*Find first word */

    while (word != NULL) {
      if (strncasecmp(word, "exit", 150) == 0) {
        exit(0);
      }
      // Some responses based on the keywords
      if (ht_get(hashtable, word) != NULL) {
        printf("\n$ (chatbot) %s\n", ht_get(hashtable, word));
      } else {
        printf("\n$ (chatbot) %s\n", "Sorry, I don't know what to say about that" );
      }
      word = strtok(NULL, SEPCHARS);
    }
  }
}

File: zig/src/chatbot.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

pub const LineLength = 80;
pub const SeparatorChars = " .,\"\n";

/// Entry in the hash table
pub const Entry = struct {
    key: []const u8,
    value: []const u8,
    next: ?*Entry,
};

/// Hash table implementation for storing responses
pub const HashTable = struct {
    size: usize,
    table: []?*Entry,
    allocator: Allocator,

    /// Create a new hash table
    pub fn create(allocator: Allocator, size: usize) !*HashTable {
        if (size < 1) return error.InvalidSize;

        const ht = try allocator.create(HashTable);
        ht.table = try allocator.alloc(?*Entry, size);
        @memset(ht.table, null);
        ht.size = size;
        ht.allocator = allocator;

        return ht;
    }

    /// Hash a string key
    fn hash(self: *const HashTable, key: []const u8) usize {
        var hashval: u64 = 0;
        for (key) |char| {
            hashval = (hashval << 8) +% @as(u64, char);
        }
        return hashval % self.size;
    }

    /// Insert or update a key-value pair
    pub fn set(self: *HashTable, key: []const u8, value: []const u8) !void {
        const bin = self.hash(key);
        var next = self.table[bin];
        var last: ?*Entry = null;

        // Find the correct position in the chain
        while (next) |current| {
            const cmp = std.mem.order(u8, current.key, key);
            if (cmp == .gt) {
                break;
            }

            if (cmp == .eq) {
                // Key exists, update value
                self.allocator.free(current.value);
                current.value = try self.allocator.dupe(u8, value);
                return;
            }

            last = current;
            next = current.next;
        }

        // Key doesn't exist, create new entry
        const newpair = try self.allocator.create(Entry);
        newpair.key = try self.allocator.dupe(u8, key);
        newpair.value = try self.allocator.dupe(u8, value);
        newpair.next = next;

        if (last) |l| {
            l.next = newpair;
        } else {
            self.table[bin] = newpair;
        }
    }

    /// Retrieve a value by key
    pub fn get(self: *const HashTable, key: []const u8) ?[]const u8 {
        const bin = self.hash(key);
        var pair = self.table[bin];

        while (pair) |current| {
            const cmp = std.mem.order(u8, current.key, key);
            if (cmp == .gt) {
                break;
            }

            if (cmp == .eq) {
                return current.value;
            }

            pair = current.next;
        }

        return null;
    }

    /// Cleanup hash table resources
    pub fn destroy(self: *HashTable) void {
        for (self.table) |entry_opt| {
            var entry = entry_opt;
            while (entry) |current| {
                const next = current.next;
                self.allocator.free(current.key);
                self.allocator.free(current.value);
                self.allocator.destroy(current);
                entry = next;
            }
        }
        self.allocator.free(self.table);
        self.allocator.destroy(self);
    }
};

test "create_hashtable" {
    var gpa: std.heap.DebugAllocator(.{}) = .init;
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const ht = try HashTable.create(allocator, 100);
    defer ht.destroy();

    try std.testing.expectEqual(ht.size, 100);
}

test "set_and_get" {
    var gpa: std.heap.DebugAllocator(.{}) = .init;
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var ht = try HashTable.create(allocator, 100);
    defer ht.destroy();

    try ht.set("key", "value");
    const result = ht.get("key");

    try std.testing.expect(result != null);
    try std.testing.expectEqualSlices(u8, result.?, "value");
}

test "get_nonexistent_key" {
    var gpa: std.heap.DebugAllocator(.{}) = .init;
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var ht = try HashTable.create(allocator, 100);
    defer ht.destroy();

    const result = ht.get("nonexistent");

    try std.testing.expectEqual(result, null);
}

test "update_existing_key" {
    var gpa: std.heap.DebugAllocator(.{}) = .init;
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var ht = try HashTable.create(allocator, 100);
    defer ht.destroy();

    try ht.set("key", "value1");
    try ht.set("key", "value2");

    const result = ht.get("key");
    try std.testing.expect(result != null);
    try std.testing.expectEqualSlices(u8, result.?, "value2");
}

test "multiple_entries" {
    var gpa: std.heap.DebugAllocator(.{}) = .init;
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var ht = try HashTable.create(allocator, 100);
    defer ht.destroy();

    try ht.set("hi", "hello");
    try ht.set("hey", "hello");
    try ht.set("python", "Yo, I love Python");

    try std.testing.expect(ht.get("hi") != null);
    try std.testing.expectEqualSlices(u8, ht.get("hi").?, "hello");

    try std.testing.expect(ht.get("python") != null);
    try std.testing.expectEqualSlices(u8, ht.get("python").?, "Yo, I love Python");

    try std.testing.expect(ht.get("missing") == null);
}
